File: frontend/src/api/match.ts
import { http, type LoadingAxiosRequestConfig } from './http'

/** 用户简要信息（匹配到的对方） */
export interface MatchPeer {
  id: number
  nickname: string
  avatar_url: string | null
  school_id: number
  /** @deprecated 已弃用，改用 age */
  grade: string | null
  /** 年龄（从生日动态计算，13-18；可能为 null） */
  age: number | null
  gender: 'male' | 'female' | 'unknown'
}

/** 匹配会话 */
export interface MatchSession {
  id: number
  user_a: number
  user_b: number
  status: 'active' | 'ended' | 'expired'
  expires_at: string | null
  ended_at: string | null
  mutual_follow: boolean
  created_at: string | null
  peer: MatchPeer | null
}

/** 入队响应 */
export interface EnqueueResult {
  status: 'waiting' | 'matched'
  queue_id?: number
  session?: MatchSession
}

/** 临时聊天消息 */
export interface MatchMessage {
  id: number
  session_id: number
  sender_id: number
  content: string
  created_at: string | null
}

/** 历史会话列表项 */
export interface MatchHistoryItem {
  id: number
  user_a: number
  user_b: number
  status: 'active' | 'ended' | 'expired'
  expires_at: string | null
  ended_at: string | null
  mutual_follow: boolean
  created_at: string | null
  peer: MatchPeer | null
}

export interface MatchHistory {
  items: MatchHistoryItem[]
  total: number
  page: number
  page_size: number
}

/** 入队 payload */
export interface MatchEnqueuePayload {
  /** @deprecated 已弃用，改用 age_min/age_max */
  grades?: string[]
  school_ids?: number[]
  tags?: string[]
  tag_required?: string[]
  target_gender?: 'male' | 'female' | 'any'
  /** 期望对方年龄下限（13-18，None 表示不限） */
  age_min?: number | null
  /** 期望对方年龄上限（13-18，None 表示不限） */
  age_max?: number | null
}

export function enqueueMatch(payload: MatchEnqueuePayload) {
  return http.post<unknown, { data: { code: number; msg: string; data: EnqueueResult } }>('/match/queue', payload)
}

export function cancelMatch() {
  return http.post<unknown, { data: { code: number; msg: string; data: { ok: boolean } } }>('/match/cancel')
}

export function getActiveSession(config: LoadingAxiosRequestConfig = {}) {
  return http.get<unknown, { data: { code: number; msg: string; data: MatchSession | null } }>('/match/active-session', config)
}

export function listSessionMessages(sessionId: number, config: LoadingAxiosRequestConfig = {}) {
  return http.get<unknown, { data: { code: number; msg: string; data: MatchMessage[] } }>(`/match/sessions/${sessionId}/messages`, config)
}

export function listMatchHistory(page = 1, pageSize = 20) {
  return http.get<unknown, { data: { code: number; msg: string; data: MatchHistory } }>('/match/history', {
    params: { page, page_size: pageSize },
  })
}

File: backend/app/api/routes/match.py
"""实时匹配 API 路由。

- POST /match/queue：加入匹配队列
- POST /match/cancel：取消等待中的匹配
- GET  /match/active-session：查询当前活动会话（如有）
- GET  /match/sessions/{id}/messages：查询会话历史消息
- GET  /match/history：查询历史匹配会话列表
"""
from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.api.deps import current_user, verified_user
from app.core.database import get_db
from app.models import User
from app.schemas.common import ok
from app.services import match_service

router = APIRouter(prefix="/match", tags=["match"])


class MatchEnqueuePayload(BaseModel):
    grades: list[str] = Field(default_factory=list)
    school_ids: list[int] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)               # 尽量有（软排序）
    tag_required: list[str] = Field(default_factory=list)       # 必须有（硬过滤）
    target_gender: str = Field(default="any", pattern="^(male|female|any)$")
    # 年龄系统：期望对方年龄范围（None 表示不限）
    age_min: int | None = Field(default=None, ge=13, le=18)
    age_max: int | None = Field(default=None, ge=13, le=18)


@router.post("/queue")
def enqueue_match(
    payload: MatchEnqueuePayload,
    db: Session = Depends(get_db),
    user: User = Depends(verified_user),
) -> dict:
    """加入匹配队列：需要 verified 状态（已填邀请码）。"""
    return ok(match_service.enqueue_match(
        db,
        user,
        grades=payload.grades,
        school_ids=payload.school_ids,
        tags=payload.tags,
        tag_required=payload.tag_required,
        target_gender=payload.target_gender,
        age_min=payload.age_min,
        age_max=payload.age_max,
    ))


@router.post("/cancel")
def cancel_match(
    db: Session = Depends(get_db),
    user: User = Depends(current_user),
) -> dict:
    """取消等待中的匹配。"""
    return ok(match_service.cancel_match(db, user))


@router.get("/active-session")
def get_active_session(
    db: Session = Depends(get_db),
    user: User = Depends(current_user),
) -> dict:
    """查询当前活动会话（如有）。"""
    return ok(match_service.get_active_session(db, user))


@router.get("/sessions/{session_id}/messages")
def list_session_messages(
    session_id: int,
    db: Session = Depends(get_db),
    user: User = Depends(current_user),
) -> dict:
    """查询会话历史消息。"""
    return ok(match_service.list_session_messages(db, session_id, user))


@router.get("/history")
def match_history(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    db: Session = Depends(get_db),
    user: User = Depends(current_user),
) -> dict:
    """查询历史匹配会话列表。"""
    return ok(match_service.list_my_match_history(db, user, page, page_size))
